subsys/bluetooth/controller/rng_driver.c

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <entropy.h>
+#include <soc.h>
+#include <device.h>
+#include <irq.h>
+#include <kernel.h>
+#include <kernel_includes.h>
+#include <ble_controller.h>
+#include <ble_controller_soc.h>
+
+#define DEV_DATA(dev) \
+	((struct rng_driver_data *)(dev)->driver_data)
+
+/* The RNG driver data. */
+struct rng_driver_data {
+	struct k_sem sem_sync; /* Used to fill the pending client buffer with new
+	                        * random values after RNG IRQ is triggered. */
+};
+
+static struct rng_driver_data rng_data;
+
+static int rng_driver_get_entropy(struct device *dev, u8_t *buf, u16_t len)
+{
+	__ASSERT_NO_MSG(&rng_data == DEV_DATA(dev));
+
+	uint8_t pool_capacity;
+	if (ble_controller_rand_pool_capacity_get(&pool_capacity) != 0) {
+		return -EINVAL;
+	}
+
+	u16_t bread;
+	while (len > 0) {
+		bread = min(len, (u16_t)pool_capacity);
+
+		__ASSERT_NO_MSG(bread <= (uint8_t)(-1));
+
+		while (ble_controller_rand_vector_get(buf, (uint8_t)bread) != 0) {
+			/* Put the thread on wait until next interrupt to get more
+			 * random values. */
+			k_sem_take(&rng_data.sem_sync, K_FOREVER);
+		}
+
+		buf += bread;
+		len -= bread;
+	}
+
+	return 0;
+}
+
+static int rng_driver_get_entropy_isr(struct device *dev, u8_t *buf, u16_t len,
+				      u32_t flags)
+{
+	__ASSERT_NO_MSG(&rng_data == DEV_DATA(dev));
+
+	u16_t bread;
+	uint8_t pool_capacity;
+	if (ble_controller_rand_pool_capacity_get(&pool_capacity) != 0) {
+		return -EINVAL;
+	}
+
+	if (likely((flags & ENTROPY_BUSYWAIT) == 0)) {
+		bread = min(len, (u16_t)pool_capacity);
+
+		__ASSERT_NO_MSG(bread <= (uint8_t)(-1));
+
+		int32_t errcode = ble_controller_rand_vector_get(buf, (uint8_t) bread);
+		return errcode == 0 ? bread : -EINVAL;
+	}
+
+	u16_t bleft = len;
+	while (bleft > 0) {
```

Comment:
Merge the while block and the if block above?

Reply:
These are 2 different behaviors